### ISSUE TITLE
ORA Unable to Load after setting File upload response to Required

### ISSUE THREAD

**Issue by geckiss** [overhangio/tutor#382]
**Bug description**
I have created an ORA Unit in my course. With default setting it loads correctly, but after changing File Upload Response to 'Required' and saving, the first step of ORA, 'Your response', is red with message `Unable to load` and spinning wheel.
This message appears on the bottom of the screen:
```
Studio's having trouble saving your work

The Studio servers encountered an error
```

**How to reproduce**
Create an ORA problem in Unit of your course. Click on edit in ORA toolbar, go to settings in top right of the modal and in File Upload Response dropdown, set the value to 'Required'.

**Environment**
Ubuntu 18.04.5 LTS
Tutor 10.1.0

**Additional context**
I got this from logs:
```
cms_1             | 2020-11-10 18:07:47,040 ERROR 30 [django.request] [user 22] log.py:228 - Internal Server Error: /preview/xblock/block-v1:TEST+Test001+2020+type@openassessment+block@65d584a986c94f3fa0
214a9b0e41d2ef/handler/render_submission                                                                                                                                                                   
cms_1             | Traceback (most recent call last):                                                                                                                                                     
cms_1             |   File "/openedx/venv/lib/python3.5/site-packages/django/core/handlers/exception.py", line 34, in inner                                                                                
cms_1             |     response = get_response(request)                                                                                                                                                   
cms_1             |   File "/openedx/venv/lib/python3.5/site-packages/django/core/handlers/base.py", line 115, in _get_response                                                                            
cms_1             |     response = self.process_exception_by_middleware(e, request)                                                                                                                        
cms_1             |   File "/openedx/venv/lib/python3.5/site-packages/django/core/handlers/base.py", line 113, in _get_response                                                                            
cms_1             |     response = wrapped_callback(request, *callback_args, **callback_kwargs)                                                                                                            
cms_1             |   File "/opt/pyenv/versions/3.5.9/lib/python3.5/contextlib.py", line 30, in inner                                                                                                      
cms_1             |     return func(*args, **kwds)                                                                                                                                                         
cms_1             |   File "/openedx/venv/lib/python3.5/site-packages/django/contrib/auth/decorators.py", line 21, in _wrapped_view                                                                        
cms_1             |     return view_func(request, *args, **kwargs)                                                                                                                                         
cms_1             |   File "/openedx/venv/lib/python3.5/site-packages/django/views/decorators/clickjacking.py", line 50, in wrapped_view                                                                   
cms_1             |     resp = view_func(*args, **kwargs)                                                                                                                                                  
cms_1             |   File "/openedx/edx-platform/cms/djangoapps/contentstore/views/preview.py", line 73, in preview_handler                                                                               
cms_1             |     resp = instance.handle(handler, req, suffix)                                                                                                                                       
cms_1             |   File "/openedx/venv/lib/python3.5/site-packages/xblock/mixins.py", line 89, in handle                                                                                                
cms_1             |     return self.runtime.handle(self, handler_name, request, suffix)                                                                                                                    
cms_1             |   File "/openedx/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1453, in handle
cms_1             |     return super(MetricsMixin, self).handle(block, handler_name, request, suffix=suffix)                                                                                             
cms_1             |   File "/openedx/venv/lib/python3.5/site-packages/xblock/runtime.py", line 1063, in handle
cms_1             |     results = handler(request, suffix)                                                                                                                                                 
cms_1             |   File "/openedx/venv/lib/python3.5/site-packages/openassessment/xblock/submission_mixin.py", line 675, in render_submission                                                           
cms_1             |     path, context = self.submission_path_and_context()                                                                                                                                 cms_1             |   File "/openedx/venv/lib/python3.5/site-packages/openassessment/xblock/submission_mixin.py", line 715, in submission_path_and_context                                                 
cms_1             |     context['file_urls'] = self.file_manager.file_descriptor_tuples(include_deleted=True)                                                                                              cms_1             |   File "/openedx/venv/lib/python3.5/site-packages/openassessment/fileupload/api.py", line 311, in file_descriptor_tuples                                                               
cms_1             |     for upload in self.get_uploads(include_deleted=include_deleted):                                                                                                                   cms_1             |   File "/openedx/venv/lib/python3.5/site-packages/openassessment/fileupload/api.py", line 243, in get_uploads                                                                          cms_1             |     user_uploads = self._file_uploads_from_list_fields(descriptions, names, sizes, include_deleted=include_deleted)                                                                    
cms_1             |   File "/openedx/venv/lib/python3.5/site-packages/openassessment/fileupload/api.py", line 502, in _file_uploads_from_list_fields
cms_1             |     return self._descriptionless_uploads()
cms_1             |   File "/openedx/venv/lib/python3.5/site-packages/openassessment/fileupload/api.py", line 536, in _descriptionless_uploads
cms_1             |     download_url = get_download_url(file_key)
cms_1             |   File "/openedx/venv/lib/python3.5/site-packages/openassessment/fileupload/api.py", line 38, in get_download_url
cms_1             |     url = backends.get_backend().get_download_url(key)
cms_1             |   File "/openedx/venv/lib/python3.5/site-packages/openassessment/fileupload/backends/filesystem.py", line 44, in get_download_url
cms_1             |     make_download_url_available(self._get_key_name(key), self.DOWNLOAD_URL_TIMEOUT)
cms_1             |   File "/openedx/venv/lib/python3.5/site-packages/openassessment/fileupload/backends/filesystem.py", line 95, in make_download_url_available
cms_1             |     get_cache().set(
cms_1             |   File "/openedx/venv/lib/python3.5/site-packages/openassessment/fileupload/backends/filesystem.py", line 70, in get_cache
cms_1             |     return django.core.cache.caches[cache_name]
cms_1             |   File "/openedx/venv/lib/python3.5/site-packages/django/core/cache/__init__.py", line 76, in __getitem__
cms_1             |     "Could not find config for '%s' in settings.CACHES" % alias
cms_1             | django.core.cache.backends.base.InvalidCacheBackendError: Could not find config for 'ora2-storage' in settings.CACHES
cms_1             | REDACTED_IP - - [10/Nov/2020:18:07:47 +0000] "POST /preview/xblock/block-v1:TEST+Test001+2020+type@openassessment+block@65d584a986c94f3fa0214a9b0e41d2ef/handler/render_submission HT
TP/1.0" 500 52 "https://redacted.domain/container/block-v1:TEST+Test001+2020+type@vertical+block@d12f04a64f914f0d8492e775c6565df2" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:82.0) Gecko/20100101 
Firefox/82.0"
```

After searching edX docs, I found https://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/configuration/ora2/ora2_uploads.html - by default, ORA stores files in Amazon S3, which I doubt is configured in Tutor(e.g. which credentials would you use?). And as you can see from logs, Django could not found config for 'ora2-storage'. 

Accroding to the link I posted above, I know how to configure ORA to store files locally, but I can't find the files mentioned, 'base.py' and the init file. 

If I exec into the cms_1 container I can see the files there. But editing them inside is not suitable since containers are ephemeral and the path is not mounted. If my container crashes, I'd have to code it again. Meh.

My question is - is there any way I can configure ORA in Tutor, without actually modifying the Dockerfile(inserting volume for the .py files) ?

**Comment by regisb** [comment]
I find this issue especially interesting because I was the one who implemented local file storage in ora2, as one of my [very first](https://github.com/edx/edx-ora2/pull/640) Open edX pull requests :innocent: 

With Tutor, the LMS includes an `ora2-storage` cache entry. The CMS doesn't because I thought there was no reason to upload actual evaluation files from within the CMS. This is what is causing the issue. At this point, the openassessment unit is correctly saved in the CMS, but it does not render correctly because it cannot access the ora2-storage cache. If you click "view live version" or "preview" you should be able to view the properly displayed unit in the LMS.

So this should not cause "real" issues in production. Nonetheless, I understand that it is annoying and confusing. I attempted to add the "ora2-storage" cache to the CMS configuration, but then the display of the ora2 unit in the CMS causes another issue, this one due to a missing url.

I believe the openassessment xblock will be considerably overhauled in Koa, which is scheduled for release in a couple weeks. I propose we close this for now and re-open if this issue still occurs in Koa. What do you thing @geckiss?

**Comment by geckiss** [comment]
Hi,
sorry for the late response. 

I understand the problem now, thanks for your reply. 

We can close this and wait for Koa, hopefully it will solve my problem :)

**Comment by regisb** [comment]
Unfortunately I just reproduced the issue on the newly-created open-release/koa.test01 branch which uses ora2==2.11.5  :-( We should check whether the problem also occurs on a native installation. If it is the case, then it means that this is an upstream issue.

**Comment by regisb** [comment]
ping @morenol: would you mind taking a look at this issue? tl;dr: in Tutor, the openassessment xblock displays an error when "File Upload Response" is "Required" and the xblock is configured for local file storage. Do you know if this error also occurs in the devstack or in the native installation? I don't mean to assign this issue to you, don't worry :) But I know that you've been working on ora2 a lot lately.

Thanks!

**Comment by morenol** [comment]
I confirm that I in the devstack this is happening when this setting is in place `ORA2_FILEUPLOAD_BACKEND = 'filesystem'`

![image](https://user-images.githubusercontent.com/22335041/99849707-68c53780-2b52-11eb-838d-c606617d0174.png)


```python
edx.devstack.lms                  | Traceback (most recent call last):
edx.devstack.lms                  |   File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/openassessment/xblock/submission_mixin.py", line 447, in upload_url
edx.devstack.lms                  |     url = file_upload_api.get_upload_url(key, content_type)
edx.devstack.lms                  |   File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/openassessment/fileupload/api.py", line 29, in get_upload_url
edx.devstack.lms                  |     return backends.get_backend().get_upload_url(key, content_type)
edx.devstack.lms                  |   File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/openassessment/fileupload/backends/filesystem.py", line 40, in get_upload_url
edx.devstack.lms                  |     make_upload_url_available(self._get_key_name(key), self.UPLOAD_URL_TIMEOUT)
edx.devstack.lms                  |   File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/openassessment/fileupload/backends/filesystem.py", line 81, in make_upload_url_available
edx.devstack.lms                  |     get_cache().set(
edx.devstack.lms                  |   File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/openassessment/fileupload/backends/filesystem.py", line 69, in get_cache
edx.devstack.lms                  |     raise exceptions.FileUploadInternalError("Undefined cache backend for file upload")
edx.devstack.lms                  | openassessment.fileupload.exceptions.FileUploadInternalError: Undefined cache backend for file upload
```

**Comment by morenol** [comment]
My bad, it works. 

I needed these settings too, that I didn't have defined:

```python
ORA2_FILEUPLOAD_CACHE_NAME = 'ora2storage'
ORA2_FILEUPLOAD_ROOT = '/edx/var/ora2/'
```

And I also a ora2storage entry to the CACHES configuration on lms.yml. I guess that tutor already has that configured, right? @regisb

**Comment by morenol** [comment]
Update again, it seems to me that it is not working fine in the devstack. 

When I access to a new component in the lms it is displaying like if I already uploaded files

![image](https://user-images.githubusercontent.com/22335041/99855870-29044d00-2b5e-11eb-9794-ff43ea58ecd3.png)

**Comment by regisb** [comment]
> When I access to a new component in the lms it is displaying like if I already uploaded files

This is actually a different issue. Interestingly enough, a PR was just opened on Openfun's repo to address this very issue: https://github.com/openfun/openedx-docker/pull/266

**Comment by regisb** [comment]
@morenol Are you quite sure that the Ora2 unit works in the CMS after adding the proposed Cache settings? I just launched a devstack and it fails with the `NoReverseMatch` error [I mentioned above](https://github.com/overhangio/tutor/issues/382#issuecomment-725331543) -- which makes sense because the fileupload urls are present in `lms/urls.py` but not 'cms/urls.py`: 

```
urlpatterns += [
    url(r'^openassessment/fileupload/', include('openassessment.fileupload.urls')),
]
```

**Comment by morenol** [comment]
I only tested that in the LMS, and yes it should not work in the CMS for that reason. But I think that it is not related to the backend, it also should fail with the other backends

**Comment by regisb** [comment]
Right @morenol, makes sense. I'll add an upstream issue then. Thanks for your input!

**Comment by regisb** [comment]
Here is the upstream issue: https://openedx.atlassian.net/jira/software/projects/BTR/boards/641?selectedIssue=BTR-38

**Comment by regisb** [comment]
This is now resolved in Koa (v11.0.0).
